wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,327 @@
+/**
+	DamageModify
+	by Trokkin
+
+	API:
+
+	// Pretty much '(Damage d) -> ...'
+	interface ForDamageCallback
+		run(Damage d)
+
+	// these are called before the damage is dealt
+	// can be called multiple amount of times or not called at all
+	addDamageModifier(ForDamageCallback cb)
+	removeDamageModifier(ForDamageCallback cb)
+	
+	// these are called after the damage is dealt
+	// are called strictly once per damage instance
+	addDamageListener(ForDamageCallback cb)
+	removeDamageListener(ForDamageCallback cb)
+
+	// a short for 'new Damage(this, target, amount, isSpell, id)..modify()..deal()'
+	unit.damage(unit target, real amount, OPTIONAL bool isSpell = true, int id = 0)
+
+	// a short for '= new Damage(this, target, amount, isSpell, id)..saveInstance()..modify()..deal()'
+	unit.returnDamage(unit target, real amount, bool isSpell, int id) -> Damage
+
+	class Damage
+		
+		// Basic damage event info
+		unit source
+		unit target
+		bool spell
+		real initial
+
+		// .id is used for any unique iterations
+		integer id
+		// .amount is used to manipulate damage amount
+		// like, addDamageModifier(d -> d.amount += d.initial * 1.00) adds +100% multiplier onto damage
+		real amount
+
+		// if .saveTarget was called then .target will survive from this damage no matter what
+		saveTarget()
+
+		// after damage has happened the Damage object is destroyed if saveInstance() wasn't called
+		saveInstance()
+		// releaseInstance is used to destroy Damage objects. Use this instead of 'destroy'!
+		releaseInstance()
+
+		// Some obvious wrappers
+		addAmount(real amount)
+		addMultiplier(real mult)
+
+		// .modify() evaluates all damageModifiers on itself
+		// is automatically called on non-triggered damage once
+		// isn't called on manual construction, so user can manually decide how many iterations a damage should have, from 0 to infinity
```

Comment:
I don't understand this. It sounds like `modify` is the function whose side-effect is to "apply" all attached damage modifiers, but not yet deal damage, is that right?

Reply:
Right. The main purpose is to allow intervention at any stage when dealing damage from code.
For example, you can additionally modify the damage before it is dealt after modify method is called.

Reply:
I find that a bit weird. Maybe could use a diagram in documentation